### PR TITLE
[FSDP, VLM] feat: add vlm training for FSDP

### PR DESCRIPTION
**Pull request by nanjiangwill**
Goal: Support VLM training on slime with FSDP

TODO
- [X] model support
- - [X] add support for `class AutoModelForVision2Seq/AutoModelForImageTextToText`
- - [X] support `Qwen2.5-VL/Qwen3-VL`(we just use default hf) new-feature TBD  @windowsXP-beta
- [ ] support dataset - refer to examples in `examples`
- - [X] for single turn -> [geo3k](https://huggingface.co/datasets/hiyouga/geometry3k), how verl did it https://github.com/volcengine/verl/pull/2014 @coding-famer @nanjiangwill 
- - [ ] for multi turn -> [deepeyes](https://arxiv.org/abs/2505.14362), how verl did it https://github.com/volcengine/verl/pull/2398
- [X] single turn VLM training
- - [X] support image in prompt/data, image processing, data pipeline @coding-famer @nanjiangwill 
- - [X] dataset properly handle multimodal info @coding-famer 
- - [X] rollout properly handle multimodal info @coding-famer @nanjiangwill 
- - [X] training(fsdp) properly handle multimodal info @nanjiangwill [this](https://github.com/volcengine/verl/blob/main/verl/workers/actor/dp_actor.py#L117)
- - [X] able to train on geo3k
- [ ] multi turn VLM training @minleminzui
- - [X] data pipeline for new images generated during the rollout stage

### CODE REVIEW

File: slime/rollout/sglang_rollout.py [hunk 1]
```diff
@@ -144,6 +155,11 @@ async def generate(args: Namespace, sample: Sample, sampling_params: dict[str, A
         if not sample.tokens:  # Initialize sample.tokens for the first turn
             sample.tokens = prompt_token_ids
 
+            # Process images for training (like tokenization for images)
+            if images_for_training and state.processor is not None:
+                processed = state.processor(images=images_for_training, return_tensors="pt")
+                sample.pixel_values = processed["pixel_values"]
```
Comment (zhuzilin) [inline slime/rollout/sglang_rollout.py]: we shouldn't try to pass the pixel values from sglang to megatron, but instead maybe re-process the image from the training side is better.

**Comment by coding-famer** [comment]
Hi @nanjiangwill, glad to see this important feature! I'd love to help with this PR if you're open to collaborating. I can help with supporting qwen-vl models and geo3k training example. Please LMK if you are happy with collaborating!

**Comment by nanjiangwill** [comment]
> Hi @nanjiangwill, glad to see this important feature! I'd love to help with this PR if you're open to collaborating. I can help with supporting qwen-vl models and geo3k training example. Please LMK if you are happy with collaborating!

hi @coding-famer, that will be amazing! can i have ur email/contact number so i can reach out to you?

**Comment by coding-famer** [comment]
> > Hi @nanjiangwill, glad to see this important feature! I'd love to help with this PR if you're open to collaborating. I can help with supporting qwen-vl models and geo3k training example. Please LMK if you are happy with collaborating!
> 
> hi @coding-famer, that will be amazing! can i have ur email/contact number so i can reach out to you?

Have sent you an email!

**Comment by dongyuanjushi** [comment]
Hi @nanjiangwill and @coding-famer, I'd love to help with this PR about the multi-turn part if you are open to this.

**Comment by nanjiangwill** [comment]
> Hi @nanjiangwill and @coding-famer, I'd love to help with this PR about the multi-turn part if you are open to this.

heyy thanks for reaching out! can i have your email?

**Comment by dongyuanjushi** [comment]
> > Hi @nanjiangwill and @coding-famer, I'd love to help with this PR about the multi-turn part if you are open to this.
> 
> heyy thanks for reaching out! can i have your email?

Have sent the email!

**Comment by zhaochenyang20** [comment]
我的神 @nanjiangwill 😭

**Comment by zhaochenyang20** [comment]
牛逼！

**Comment by aJupyter** [comment]
请问大概什么时候才会合并呀

**Comment by zhaochenyang20** [comment]
> 请问大概什么时候才会合并呀

肯定会早于星际之门集群建设好

File: slime/backends/sglang_utils/sglang_engine.py [hunk 2]
```diff
@@ -31,6 +31,7 @@ def get_base_gpu_id(args, rank):
 
 
 def launch_server_process(server_args: ServerArgs) -> multiprocessing.Process:
+    multiprocessing.set_start_method("spawn", force=True)
```
Comment (nanjiangwill) [inline slime/backends/sglang_utils/sglang_engine.py:34]: without doing this there will be a lot of mm_processor warning from sglang, would be curious if there is a better way to resolve the issue @zhuzilin

**Comment by tellyoung** [comment]
Could the code for multimodal processing be merged as soon as possible? The support for multi-turn is a separate module and can be merged later. Otherwise, this will affect the issues based on the multimodal model and merge code.

**Comment by jhinpan** [comment]
> Could the code for multimodal processing be merged as soon as possible? The support for multi-turn is a separate module and can be merged later. Otherwise, this will affect the issues based on the multimodal model and merge code.

This PR will be merge in the next few days. Pls just give us some time to do final check and review.

File: slime/backends/sglang_utils/sglang_engine.py (hunk 2 above)
Comment (jhinpan) [reply slime/backends/sglang_utils/sglang_engine.py:34]: p.s: Img of mm_process warning:
image

Comment (jhinpan) [reply slime/backends/sglang_utils/sglang_engine.py:34]: Confirmed by @zhuzilin: "In sglang, the engine part already has this set by default (https://github.com/sgl-project/sglang/blob/303cc957e62384044dfa8e52d7d8af8abe12f0ac/python/sglang/srt/entrypoints/engine.py#L747), so it should be fine to add it."

**Comment by zhaochenyang20** [comment]
🈚️敌！

**Comment by jhinpan** [comment]
We ran three experiments comparing different reward model configurations on 8*H100:
  1. geo3k RM with tol=0.05 (original)
  2. geo3k RM with tol=0.0 (strict matching)
  3. default math RM

  Results showed all three configurations perform similarly, indicating:
  - The tolerance parameter has minimal impact on training outcomes
  - The geo3k-specific reward model provides no advantage over the default math RM
  
 I will later remove them. Just leave a note here to mark:
 
image

File: slime/backends/fsdp_utils/data_packing.py [hunk 3]
```diff
@@ -95,6 +111,8 @@ def pack_sequences(
                 "rollout_log_probs": torch.tensor(
                     flat_rollout_log_probs, dtype=torch.float32, device=torch.cuda.current_device()
                 ),
+                "multimodal_inputs": multimodal_data,
+                "multimodal_num_items": multimodal_num_items,
```
Comment (zhuzilin) [inline slime/backends/fsdp_utils/data_packing.py]: here we need sth like:
```python
packed_batch = {
    "tokens": ...,
}
if multimodal_inputs:
    for key, mm_tensor in multimodal_inputs[i].items():
        ...
    packed_batch.extend({
        "multimodal_inputs": multimodal_data,
        "multimodal_num_items": multimodal_num_items,
    })
result.append(packed_batch)
```

File: slime/ray/rollout.py [hunk 4]
```diff
@@ -194,7 +194,7 @@ def _post_process_rewards(self, samples: list[Sample] | list[list[Sample]]):
             and self.args.rewards_normalization
         ):
             # group norm
-            rewards = torch.tensor(raw_rewards, dtype=torch.float)
+            rewards = torch.tensor(raw_rewards, dtype=torch.float16)
```
Comment (zhuzilin) [inline slime/ray/rollout.py]: hmm.. it seems better to move this type conversion into the custom reward model? otherwise, it may influence the other users who are using dense rm for rlhf.

File: slime/backends/fsdp_utils/data_packing.py (hunk 3 above)
Comment (jhinpan) [reply slime/backends/fsdp_utils/data_packing.py]: used .update() instead of .extend() since dictionaries use update to merge key-value pairs.

File: slime/ray/rollout.py (hunk 4 above)
Comment (jhinpan) [reply slime/ray/rollout.py]: move float16 specifically to notes in geo3k_vlm example

**Comment by simondong1** [comment]
Hi @nanjiangwill, this is awesome! I am also working on RL with VLMs. Would love to contribute and collaborate on further works! my email is simondong0919 at gmail.com love to get in touch

**Comment by adol001** [comment]
@nanjiangwill @zhuzilin 

It looks like this commit changed the behavior of the apply_chat_template parameter. As a result, this parameter becomes useless during dataset construction: instead of applying the chat template, the data from prompt_key is fed directly into something like:

messages = [{"role": "user", "content": messages}]


In custom generate scenarios, do you realize what this can lead to?

You’re effectively replacing what used to be a str with a dict (or even a list of dicts). If downstream code doesn’t explicitly validate types, it will just get stuffed into the prompt and you end up with outputs like:

First rollout sample: ['<|im_start|>user\n[{\'role\': \'user\', \'content\':


This kind of change alters the default behavior in a non-backward-compatible way. At the very least, compatibility with existing users should be considered — and there should be warnings or clear notices. In my case, it indirectly caused training quality to degrade. I wouldn’t have noticed this logic change if it weren’t for a new training job that made the issue obvious.

**Comment by zhuzilin** [comment]
@adol001 really sorry about this, we will revert this change.

**Comment by nanjiangwill** [comment]
@adol001 sorry again about this, we have fixed the issue with #1232 and #1234
